helix-view/src/document.rs

```diff
@@ -115,6 +116,58 @@ pub struct SavePoint {
     revert: Mutex<Transaction>,
 }
 
+// An enum representing the encoding and if it contain a byte order mark
+#[derive(Debug, Clone)]
+pub enum EncodingBom {
+    UTF8,
+    NoBom(&'static Encoding),
+}
+
+impl EncodingBom {
+    fn encoding(&self) -> &'static Encoding {
+        match self {
+            EncodingBom::UTF8 => encoding::UTF_8,
+            EncodingBom::NoBom(encoding) => encoding,
+        }
+    }
+
+    // Assume that encoding have a bom, used to convert
+    // `encoding_rs` for_bom result to this enum
+    fn for_bom(encoding: &'static Encoding) -> Self {
+        if encoding == encoding::UTF_8 {
+            EncodingBom::UTF8
+        } else {
+            EncodingBom::NoBom(encoding)
+        }
+    }
+
+    // Write in buf the Bom if applicable
+    // return the number of bytes written
+    fn write_bom(&self, buf: &mut [u8; BUF_SIZE]) -> usize {
+        match self {
+            EncodingBom::UTF8 => {
+                buf[0] = 0xef;
+                buf[1] = 0xbb;
+                buf[2] = 0xbf;
+                3
+            }
+            EncodingBom::NoBom(_) => 0,
+        }
+    }
+}
+
+impl From<&'static Encoding> for EncodingBom {
+    fn from(value: &'static Encoding) -> Self {
+        EncodingBom::NoBom(value)
+    }
+}
+
+impl Default for EncodingBom {
+    fn default() -> Self {
+        EncodingBom::NoBom(encoding::UTF_8)
+    }
+}
```

Comment:
I'd move this right after the struct definition to make it more obvious

Reply:
This impl is now moved just below EncodingBom declaration.